### PR TITLE
Issue/4558 level3 make create customer call

### LINKED ISSUE #4558: [Card Present Payments] Add evel 3 data
pdfdoF-1i-p2

- [x] Add POST call to Fluxc (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2086)
- [x] Call the backend via Fluxc just before creating a payment intent. Make sure to show the loading state
- [ ] Pass customer_id from the step 2

### PR DESCRIPTION
**Pull request by kidinov**
Resolves #4558 

**Merge after https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2086**

The PR adds create customer calls just before creating a payment intent. The resulting customer_id is added to the intent. "Loading" state is already shown at the moment I perform the call.

https://user-images.githubusercontent.com/4923871/128342093-59446bdd-01bc-442d-bf2a-b2e950729a23.mp4

### How to test
There is no UI facing changes.

1. Make a payment
2. Go to the stripes dashboard
3. Find your transaction
4. Scroll down and find JSON which represents the call to create a payment intent
5. Notice customer is set

image
image

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt [hunk 1]
```diff
@@ -157,6 +151,24 @@ internal class PaymentManager(
...
+            !terminalWrapper.isInitialized() -> {
+                emit(errorMapper.mapError(errorMessage = "Reader not connected"))
+                true
+            }
+            else -> false
+        }
+
+    private suspend fun enrichPaymentInfoWithCustomerId(paymentInfo: PaymentInfo): PaymentInfo {
+        val customerId = cardReaderStore.getCustomerIdByOrderId(paymentInfo.orderId)
+        return paymentInfo.copy(customerId = customerId)
```
Comment (kidinov) [inline libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:170]: I don't really like this approach, but adding the call in the VM is also a bit off because it's more like a payments-only related functionality. Any ideas on how to make it better?

File: build.gradle [hunk 2]
```diff
@@ -94,7 +94,7 @@ tasks.register("installGitHooks", Copy) {
 }
 
 ext {
-    fluxCVersion = 'develop-0193e64889a94b44c507c6aba5b83ce0fc96bc61'
+    fluxCVersion = '2086-a4de5c92ad5fa578b4db3eccba7be0730b5a690f'
```
Comment (kidinov) [inline build.gradle]: - [x]  Change to develop after Fluxc PR is merged

File: WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt [hunk 3]
```diff
@@ -30,6 +30,10 @@ class CardReaderModule {
         payStore: WCPayStore,
         responseMapper: CapturePaymentResponseMapper
     ) = object : CardReaderStore {
+        override suspend fun getCustomerIdByOrderId(orderId: Long): String? {
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt]: Nitpick 🔍 : Technically not part of this PR - but I'd consider renaming both `getCustomerIdByOrderId` and `getConnectionToken` to `createCustomerIdByOrderId` and `fetchConnectionToken`.

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt [hunk 4]
```diff
@@ -38,14 +38,8 @@ internal class PaymentManager(
...
     suspend fun acceptPayment(paymentInfo: PaymentInfo): Flow<CardPaymentStatus> = flow {
-        if (!paymentUtils.isSupportedCurrency(paymentInfo.currency)) {
-            emit(errorMapper.mapError(errorMessage = "Unsupported currency: $paymentInfo.currency"))
-            return@flow
-        }
-        if (!terminalWrapper.isInitialized()) {
-            emit(errorMapper.mapError(errorMessage = "Reader not connected"))
-            return@flow
-        }
+        if (isValidState(paymentInfo)) return@flow
```
Comment (malinajirka) [inline libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt]: `isValidState` returns true when the state is not valid -> this feels a bit confusing. I'd consider either renaming the method or swapping true/false. Wdyt?

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt (hunk 1 above)
Comment (malinajirka) [reply libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt:170]: nitpick 🔍 : I think the approach of creating the customer from within the module and adding a method to CardReaderStore is completely fine. I'd personally not include "customerId" in PaymentInfo but send it as a separate field -> the main reason is that now the client(app) can set customerId even though it has no access to it which makes the api slightly confusing. But it's not a big deal.

File: libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt [hunk 5]
```diff
@@ -159,6 +184,27 @@ class PaymentManagerTest {
         verify(paymentErrorMapper).mapTerminalError(anyOrNull(), anyOrNull())
     }
 
+    @Test
+    fun `given status not customer id returns, when creating payment finishes, then payment info enriched`() =
```
Comment (malinajirka) [inline libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt]: Non-blocking question: `given status not customer id returns` I'm not sure I follow this part of the name, is there a mistake perhaps?

**Review by malinajirka — APPROVED** [review]
LGTM, left three non-blocking comments. Thanks for working on this ;)!

File: build.gradle (hunk 2 above)
Comment (kidinov) [reply build.gradle]: 2391e9882841caebf1254acab59d8aecb813c7b4

File: WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt (hunk 3 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt]: >  but I'd consider renaming both getCustomerIdByOrderId and getConnectionToken

Tbh I'm not sure about that, because we technically don't really care if the customer created now or it was created before. Only what the client needs is the customer id. I renamed both to `fetchCustomerIdByOrderId` and `fetchConnectionToken`. Let me know if it's ok

3f36c23ac2482bf85a23d5700fdea95e5a84ae52

File: libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt (hunk 4 above)
Comment (kidinov) [reply libs/cardreader/src/main/java/com/woocommerce/android/cardreader/internal/payments/PaymentManager.kt]: 🤦  renamed to `isInvalidState`  c46e6ac4df26856799ba63cb3f3c16e02089371b

File: libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt (hunk 5 above)
Comment (kidinov) [reply libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/payments/PaymentManagerTest.kt]: Yeah, copy-pasting. Fixed  6babb2ad687983dbde58cbc008aa352bc536797d

**Comment by kidinov** [comment]
@malinajirka 

>  I think the approach of creating the customer from within the module and adding a method to CardReaderStore is completely fine. I'd personally not include "customerId" in PaymentInfo but send it as a separate field -> the main reason is that now the client(app) can set customerId even though it has no access to it which makes the api slightly confusing. But it's not a big deal.

Agree about API part, but on other hand, `customerId` is part of the `PaymentInfo`. I going to merge it now, maybe will change it after Im coming back from the AFK

File: WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt (hunk 3 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/di/CardReaderModule.kt]: Good point, even better ;)
